libraries/chain/db_market.cpp

```diff
@@ -1166,16 +1166,42 @@ void database::pay_order( const account_object& receiver, const asset& receives,
    adjust_balance(receiver.get_id(), receives);
 }
 
-asset database::calculate_market_fee( const asset_object& trade_asset, const asset& trade_amount )
+asset database::calculate_market_fee( const asset_object& trade_asset, const asset& trade_amount, const bool is_maker)
 {
    assert( trade_asset.id == trade_amount.asset_id );
 
    if( !trade_asset.charges_market_fees() )
       return trade_asset.amount(0);
-   if( trade_asset.options.market_fee_percent == 0 )
+   // Optimization: The fee is zero if the order is a maker, and the maker fee percent is 0%
+   if( is_maker && trade_asset.options.market_fee_percent == 0 )
       return trade_asset.amount(0);
 
-   auto value = detail::calculate_percent(trade_amount.amount, trade_asset.options.market_fee_percent);
+   // BSIP81: Asset owners may specify different market fee rate for maker orders and taker orders
+   auto maint_time = get_dynamic_global_properties().next_maintenance_time;
+   bool before_core_hardfork_bsip81 = ( maint_time <= HARDFORK_BSIP_81_TIME ); // before simple maker-taker fee
```

Comment:
https://github.com/bitshares/bsips/pull/271 will make code here much simpler, only need to check whether the `taker_fee_percent` flag is valid.